### PR TITLE
ipahost: Extension to be able handle several hosts and all settings

### PR DESCRIPTION
**Pull request by t-woerner**
The ipahost management module was not able to add several hosts at once.
Addtionally there have been settings missing.

ansible_freeipa_module has been extended to provide two additional functions
that are needed to simplify the extension of the ipahost module:

    gen_add_del_lists(user_list, res_list)
    encode_certificate(cert)

gen_add_del_lists will generate the lists for the addition and removal of
members using the provided user and ipa settings.

encode_certificate will encode a certificate using base64 with also taking
FreeIPA and Python versions into account.

The missing settings in ipahost have been:

    certificate
    managedby_host
    principal
    create_keytab_[user,group,host,hostgroup]
    retrieve_keytab_[user,group,host,hostgroup]
    sshpubkey
    userclass
    auth_ind
    requires_pre_auth
    ok_as_delegate
    ok_to_auth_as_delegate

The README-host.md file has been updated to provide information about the
new settings and also the members. Also examples for the new things have
been added.

New example playbooks have been added:

    playbooks/host/add-host.yml
    playbooks/host/host-member-allow_create_keytab-absent.yml
    playbooks/host/host-member-allow_create_keytab-present.yml
    playbooks/host/host-member-allow_retrieve_keytab-absent.yml
    playbooks/host/host-member-allow_retrieve_keytab-present.yml
    playbooks/host/host-member-certificate-absent.yml
    playbooks/host/host-member-certificate-present.yml
    playbooks/host/host-member-managedby_host-absent.yml
    playbooks/host/host-member-managedby_host-present.yml
    playbooks/host/host-member-principal-absent.yml
    playbooks/host/host-member-principal-present.yml
    playbooks/host/host-present-with-allow_create_keytab.yml
    playbooks/host/host-present-with-allow_retrieve_keytab.yml
    playbooks/host/host-present-with-certificate.yml
    playbooks/host/host-present-with-managedby_host.yml
    playbooks/host/host-present-with-principal.yml
    playbooks/host/host-present-with-randompassword.yml
    playbooks/host/host-present.yml
    playbooks/host/hosts-member-certificate-absent.yml
    playbooks/host/hosts-member-certificate-present.yml
    playbooks/host/hosts-member-managedby_host-absent.yml
    playbooks/host/hosts-member-managedby_host-present.yml
    playbooks/host/hosts-member-principal-absent.yml
    playbooks/host/hosts-member-principal-present.yml
    playbooks/host/hosts-present-with-certificate.yml
    playbooks/host/hosts-present-with-managedby_host.yml
    playbooks/host/hosts-present-with-randompasswords.yml

New tests have been added for the module:

    tests/host/certificate/cert1.der
    tests/host/certificate/cert1.pem
    tests/host/certificate/cert2.der
    tests/host/certificate/cert2.pem
    tests/host/certificate/cert3.der
    tests/host/certificate/cert3.pem
    tests/host/certificate/private1.key
    tests/host/certificate/private2.key
    tests/host/certificate/private3.key
    tests/host/certificate/test_host_certificate.yml
    tests/host/certificate/test_hosts_certificate.yml
    tests/host/test_host.yml
    tests/host/test_host_allow_create_keytab.yml
    tests/host/test_host_allow_retrieve_keytab.yml
    tests/host/test_host_managedby_host.yml
    tests/host/test_host_principal.yml
    tests/host/test_host_random.yml
    tests/host/test_hosts.yml
    tests/host/test_hosts_managedby_host.yml
    tests/host/test_hosts_principal.yml

### CODE REVIEW

File: README-host.md [hunk 1]
```diff
@@ -85,28 +85,118 @@ Example playbook to create host without DNS:
...
+
+  - name: Print generated random password for host01.example.com
+    debug:
+      var: ipahost.host["host01.example.com"].randompassword
+
+  - name: Print generated random password for host02.example.com
+    debug:
+      var: ipahost.host["host02.example.com"].randompassword
+```
+Please remember that the `force` tag will also force the generation of a new random password even if the host alreay exists and if `update_password` is limited to `on_create`.
```
Comment (varunmylaraiah) [inline README-host.md]: will also force the generation of a new random password even if the host *already* exists

File: playbooks/host/host-member-certificate-absent.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+- name: Host member certificate present
```
Comment (varunmylaraiah) [inline playbooks/host/host-member-certificate-absent.yml]: - name: Host member certificate **absent**

File: playbooks/host/host-member-certificate-absent.yml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  hosts: ipaserver
+  become: true
+
+  tasks:
+  - name: Host host01.example.com member certificate present
```
Comment (varunmylaraiah) [inline playbooks/host/host-member-certificate-absent.yml]: - name: Host host01.example.com member certificate **absent**

File: README-host.md (hunk 1 above)
Comment (t-woerner) [reply README-host.md]: Fixed

File: playbooks/host/host-member-certificate-absent.yml (hunk 3 above)
Comment (t-woerner) [reply playbooks/host/host-member-certificate-absent.yml]: Fixed

File: playbooks/host/host-member-certificate-absent.yml (hunk 2 above)
Comment (t-woerner) [reply playbooks/host/host-member-certificate-absent.yml]: Fixed

File: plugins/modules/ipahost.py [hunk 4]
```diff
@@ -216,69 +454,196 @@ def gen_args(description, force, locality, location, platform, os, password,
...
+        if action == "member":
+            # certificate, managedby_host, principal,
+            # allow_create_keytab_*, allow_retrieve_keytab_*,
+            invalid = ["description", "locality", "location", "platform",
+                       "os", "password", "random", "mac_address", "sshpubkey",
+                       "userclass", "auth_ind", "requires_pre_auth",
+                       "ok_as_delegate", "ok_to_auth_as_delegate", "force",
+                       "reverse", "ip_address", "update_dns",
+                       "update_password"]
+            for x in invalid:
```
[the remark this answers is not in the document]
Comment (t-woerner) [reply plugins/modules/ipahost.py]: Yes, this is in all the modules...

**Pull request merged**